hbase-balancer/src/main/java/org/apache/hadoop/hbase/master/balancer/StochasticLoadBalancer.java

```diff
@@ -204,10 +209,11 @@ protected float getDefaultSlop() {
 
   protected List<CandidateGenerator> createCandidateGenerators() {
     List<CandidateGenerator> candidateGenerators = new ArrayList<CandidateGenerator>(4);
-    candidateGenerators.add(new RandomCandidateGenerator());
-    candidateGenerators.add(new LoadCandidateGenerator());
-    candidateGenerators.add(localityCandidateGenerator);
-    candidateGenerators.add(new RegionReplicaRackCandidateGenerator());
+    candidateGenerators.add(GeneratorType.RANDOM.ordinal(), new RandomCandidateGenerator());
```

Comment:
I prefer we just attach the generator to the enum, so here we could write a for loop instead of typing all the enums by hand.

Reply:
I tried to attach the generators. But Enum requires immutable objects. Two subclasses of Candidate Generators (FavoredNodeLoadPicker and FavoredNodeLocalityPicker )have a field for FavoredNodeManager, which can not be immutable. Given the limited number of types here, it is not too much gain to go down the rabbit hole at shown in the reverted commits.